File: src/64_Minimum_Path_Sum/main.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  int minPathSum(vector<vector<int>>& grid) {
    int rows = grid.size();
    if (!rows) return 0;
    int cols = grid[0].size();
    if (!cols) return 0;

    vector<vector<int>> costs(rows,vector<int>(cols));
    costs[0][0] = grid[0][0];

    // 初始化第一行的所有列
    for (int j = 1; j < cols; ++j) {
      //costs[0][j] += grid[0][j];//大意啦大意啦！！
      costs[0][j] = costs[0][j-1] + grid[0][j];
    }
    // 初始化第一列的所有行
    for (int i = 1; i < rows; ++i) {
      //costs[i][0] += grid[i][0];//大意啦大意啦！！
      costs[i][0] = costs[i-1][0] + grid[i][0];
    }

    for (int i = 1; i < rows; ++i) {
      for (int j = 1; j < cols; ++j) {
        costs[i][j] = min(costs[i][j-1], costs[i-1][j]) + grid[i][j];
      }
    }
    return costs.back().back();
  }
};

int main() {
  Solution gua;
  int rtn;
  vector<vector<int>> grid;

  rtn = gua.minPathSum(grid={{1,3,1},{1,5,1},{4,2,1}});
  cout << "expect value = 7, actual value = " << rtn << endl;

  return 0;
}

File: src/64_Minimum_Path_Sum/main.py

from typing import List

class Solution:
    def minPathSum1(self, grid: List[List[int]]) -> int:
        # 就是二维的DP吧，矩阵本身就是个dp数组
        m, n = len(grid), len(grid[0])
        dp = [[0 for j in range(n)] for i in range(m)]
        
        # 初始化一下第0行，第0列
        dp[0][0] = grid[0][0]
        for i in range(1,m):
            # dp[i][0] += dp[i-1][0] 大意了啊！！
            dp[i][0] = dp[i-1][0] + grid[i][0]
        
        for j in range(1,n):
            # dp[0][j] += dp[0][j-1] 大意了啊！
            dp[0][j] = dp[0][j-1] + grid[0][j]
        
        for i in range(1,m):
            for j in range(1,n):
                # 有两个来源，水平和垂直
                dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i][j]
        
        return dp[m-1][n-1]
    
    # Runtime: 88 ms, faster than 97.47%
    # Memory Usage: 15 MB, less than 92.40%
    def minPathSum(self, grid: List[List[int]]) -> int:
        # 压缩空间试一试
        m, n = len(grid), len(grid[0])
        dp = grid[0]
        for j in range(1,n):
            dp[j] += dp[j-1]
            
        for i in range(1,m):
            dp[0] += grid[i][0]
            # prev = dp[0]
            for j in range(1,n):
                # 因为不用保存左斜上方的值，所以不用保存旧值，直接存储在dp[i]里面
                #dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i][j]
                # tmp = min(dp[j], prev) + grid[i][j]
                # prev = dp[j]
                # dp[j] = tmp
                dp[j] = min(dp[j], dp[j-1]) + grid[i][j]

        return dp[-1]


if __name__ == "__main__":
    gua = Solution()
    rtn = gua.minPathSum(grid = [[1,3,1],[1,5,1],[4,2,1]])
    print("expect result = 7, actual result = %d " % rtn )

    rtn = gua.minPathSum(grid = [[1,2,3],[4,5,6]])
    print("expect result = 12, actual result = %d " % rtn )



